[libc++] LWG3382: NTTP for `pair` and `array`

JMazurkiewicz opened the pull request:
Mark LWG3382 as "Nothing To Do" and add tests.

JMazurkiewicz committed:
[libc++] LWG3382: NTTP for `pair` and `array`

ldionne commented on libcxx/test/std/containers/sequences/array/array.overview/nttp.equivalence.compile.pass.cpp:
```diff
@@ -0,0 +1,69 @@
+
+// <array>
+
+// Two values a1 and a2 of type array<T, N> are template-argument-equivalent if and only if each pair of corresponding
```
Can you add references to the LWG issue being tested here? And also in the one for `pair`.

ldionne commented on libcxx/test/std/containers/sequences/array/array.overview/nttp.verify.cpp:
```diff
@@ -0,0 +1,80 @@
+
+// <array>
+
+// array<T, N> is a structural type ([temp.param]) if T is a structural type.
```
I don't understand what you're testing here. Isn't it sufficient to check for the NTTP behavior as far as the LWG issue is concerned? I think I'm just missing the obvious here, but if you can explain that would be great!

JMazurkiewicz replied:
I'm testing whether `array` (or `pair`) is a structural type based on its template parameters. LWG-3382 states that:
* pair<T, U> is a structural type (13.2 [[temp.param]](https://wg21.link/temp.param)) if T and U are both structural types
* array<T, N> is a structural type (13.2 [[temp.param]](https://wg21.link/temp.param)) if T is a structural type

IMHO those tests address the LWG issue, although they might be an overkill.

ldionne reviewed (commented):
Thanks for tackling this! I have some comments to clarify my understanding but I think this will be good without too many changes.

JMazurkiewicz committed:
Add references to LWG-3382

Comment: https://github.com/llvm/llvm-project/pull/85811#discussion_r1581152034

JMazurkiewicz committed:
Merge remote-tracking branch 'upstream/main' into libcxx/lwg/nttp-for-pair-and-array
